initrd/bin/seal-totp

```diff
@@ -17,69 +17,77 @@ fi
 TOTP_SEALED="/tmp/secret/totp.sealed"
 
-dd \
-	if=/dev/urandom \
-	of="$TOTP_SECRET" \
-	count=1 \
-	bs=20 \
-	2>/dev/null \
-|| die "Unable to generate 20 random bytes"
+if [ "$CONFIG_TPM" = "y" ]; then
+  dd \
+    if=/dev/urandom \
+    of="$TOTP_SECRET" \
+    count=1 \
+    bs=20 \
+    2>/dev/null \
+  || die "Unable to generate 20 random bytes"
 
-secret="`base32 < $TOTP_SECRET`"
+  secret="`base32 < $TOTP_SECRET`"
 
 # Use the current values of the PCRs, which will be read
 # from the TPM as part of the sealing ("X").
 # PCR4 == 0 means that we are still in the boot process and
 # not a recovery shell.
 # should this read the storage root key?
-if ! tpm sealfile2 \
-	-if "$TOTP_SECRET" \
-	-of "$TOTP_SEALED" \
-	-hk 40000000 \
-	-ix 0 X \
-	-ix 1 X \
-	-ix 2 X \
-	-ix 3 X \
-	-ix 4 0000000000000000000000000000000000000000 \
-	-ix 7 X \
-; then
-	rm -f "$TOTP_SECRET"
-	 die "Unable to seal secret"
-fi
+  if ! tpm sealfile2 \
+    -if "$TOTP_SECRET" \
+    -of "$TOTP_SEALED" \
+    -hk 40000000 \
+    -ix 0 X \
+    -ix 1 X \
+    -ix 2 X \
+    -ix 3 X \
+    -ix 4 0000000000000000000000000000000000000000 \
+    -ix 7 X \
+  ; then
+    rm -f "$TOTP_SECRET"
+     die "Unable to seal secret"
+  fi
 
 
 # to create an nvram space we need the TPM owner password
 # and the TPM physical presence must be asserted.
 #
 # The permissions are 0 since there is nothing special
 # about the sealed file
-tpm physicalpresence -s \
-|| warn "Warning: Unable to assert physical presence"
+  tpm physicalpresence -s \
+  || warn "Warning: Unable to assert physical presence"
 
 # Try to write it without the password first, and then create
 # the NVRAM space using the owner password if it fails for some reason.
-if ! tpm nv_writevalue \
-	-in $TPM_NVRAM_SPACE \
-	-if "$TOTP_SEALED" \
-; then
-	warn 'NVRAM space does not exist? Owner password is required'
-	read -s -p "TPM Owner password: " tpm_password
-	echo
+  if ! tpm nv_writevalue \
+    -in $TPM_NVRAM_SPACE \
+    -if "$TOTP_SEALED" \
+  ; then
+    warn 'NVRAM space does not exist? Owner password is required'
+    read -s -p "TPM Owner password: " tpm_password
+    echo
 
-	tpm nv_definespace \
-		-in $TPM_NVRAM_SPACE \
-		-sz 312 \
-		-pwdo "$tpm_password" \
-		-per 0 \
-	|| die "Unable to define NVRAM space"
+    tpm nv_definespace \
+      -in $TPM_NVRAM_SPACE \
+      -sz 312 \
+      -pwdo "$tpm_password" \
+      -per 0 \
+    || die "Unable to define NVRAM space"
 
-	tpm nv_writevalue \
-		-in $TPM_NVRAM_SPACE \
-		-if "$TOTP_SEALED" \
-	|| die "Unable to write sealed secret to NVRAM"
-fi
+    tpm nv_writevalue \
+      -in $TPM_NVRAM_SPACE \
+      -if "$TOTP_SEALED" \
+    || die "Unable to write sealed secret to NVRAM"
+  fi
 
-rm -f "$TOTP_SEALED"
+  rm -f "$TOTP_SEALED"
+else
+# without a TPM, use the first 20 characters of the ROM SHA256sum 
+  echo "TPM not configured, measuring ROM directly"
+  flash.sh -s /tmp/coreboot.rom | cut -c 1-20 | tr -d '\n' > $TOTP_SECRET
```

Comment:
Why truncate the hash prematurely? Could you not get the full 120 bits of the HOTP with something like
```suggestion
  secret="`flash.sh -s /tmp/coreboot.rom  | xxd -r -p | base32 | cut -c 1-24 | tr -d '\n'`"
```
(i.e., decode the base16 string then re-encode as base32)?

Reply:
@kylerankin thoughts?